Operator manages identities - full implementation

dlapcevic opened the pull request:
**Feature: Operator Manages Cilium Identities**

CFP: https://github.com/cilium/cilium/issues/27752

**Description**

A new feature hidden behind a flag. Disabled by default.

Besides the new flag that enables the feature, there are no other user visible changes.

cilium-operator is going to manage (create and delete) Cilium Identity API objects instead of cilium-agent.

cilium-operator calculates the desired state for Cilium Identities based on watched events for Cilium Identities, Pods, Namespaces and Cilium Endpoint Slices (if enabled).

cilium-operator creates Cilium Identities on pod updates for a unique label set based on pod and namespace labels.

cilium-operator deletes Cilium Identities when their labels are not used by any pods.

cilium-agent no longer writes to Cilium Identities.

cilium-agent only watches Cilium Identities.

pod creation (cilium-cni) is blocked on getting a matching Cilium Identity from the watcher cache.


```release-note
Feature: Operator Manages Cilium Identities
A new feature hidden behind a flag. Disabled by default.
cilium-operator is going to manage (create and delete) Cilium Identity API objects instead of cilium-agent.
```

kind/feature

Signed-off-by: Dorde Lapcevic <[dordel@google.com](mailto:dordel@google.com)>

joamaki commented on operator/pkg/ciliumidentity/namespace.go:
```diff
@@ -0,0 +1,143 @@
+	defer c.nsQueue.Done(item)
+
+	nsItem := item.(queueItem)
+	err := c.reconciler.reconcileNS(nsItem.key)
```
We seem to have 3 different workqueues that are processed in parallel, which means we'll have parallel calls to the reconciler, but I don't see any concurrency control there. Are these `reconcileXYZ` calls safe to perform in parallel? 

Also do we really need 3 different work queues or could we just go with one and enqueue newtype'd keys to differentiate the work items? IMO a single sequential processing loop that pulls from the queue would be easier to follow and test.

dlapcevic replied:
The Cilium Identity (CID) controller running in cilium-operator is responsible
only for managing CID API objects. Namespace events are pushed to namespace
workqueue. Pod events are pushed to Pod workqueue. CID events are pushed to CID
workqueue. Additionally, processed items in Namespace workqueue are pushing
items to Pod workqueue, and processed items in Pod workqueue are pushing items
to CID workqueue. Mutations to CID API objects are only issued from processing
items in CID workqueue.
Diagram (added to the operator/pkg/ciliumidentity/controller.go):\
--------------------------------------------------Pod event---------CID event\
-------------------------------------------------------||-----------------||\
-------------------------------------------------------V-----------------V\
Namespace event -> Namespace workqueue -> Pod workqueue -> CID workqueue -> Mutate CID API objects

Locks are used in desiredCIDState, as well as cidCreateLock, to ensure thread safety.

Keeping separate work queues for each API object type makes it simpler to control and extend. We have a separate rate limiting for processing CID events (rateLimitCIDProcessing), which is controlled by an operator flag, so it’s adjustable. We might want to introduce similar rate limiters for other processed API objects. Adding watches and event processing for more API objects in the future will make it more difficult to maintain and extend if we keep growing the same workqueue.

joamaki commented on operator/pkg/ciliumidentity/controller.go:
```diff
@@ -0,0 +1,228 @@
+}
+
+func (c *Controller) startWorkQueues() {
+	go c.runCIDWorker()
```
We should use the worker pool for these as well as that will allow cleaning up all background workers when stopping. You might also want to look into `pkg/hive/job`, e.g. `job.OneShot` for example. Grep for example usage.

dlapcevic replied:
Done.

joamaki commented on operator/pkg/ciliumidentity/controller.go:
```diff
@@ -0,0 +1,228 @@
+	c.wp.Submit("process-cilium-identity-events", c.processCiliumIdentityEvents)
+	c.wp.Submit("process-pod-events", c.processPodEvents)
+	c.wp.Submit("process-namespace-events", c.processNamespaceEvents)
+	c.wp.Submit("process-namespace-events", c.processCiliumEndpointSliceEvents)
```
should be "process-cilium-endpointslice-events"?

dlapcevic replied:
Done.

bimmlerd committed:
Operator manages identities - full implementation

**Feature: Operator Manages Cilium Identities**

CFP: https://github.com/cilium/cilium/issues/27752

**Description**

A new feature hidden behind a flag. Disabled by default.

Besides the new flag that enables the feature, there are no other user visible changes.

cilium-operator is going to manage (create and delete) Cilium Identity API objects instead of cilium-agent.

cilium-operator calculates the desired state for Cilium Identities based on watched events for Cilium Identities, Pods, Namespaces and Cilium Endpoint Slices (if enabled).

cilium-operator creates Cilium Identities on pod updates for a unique label set based on pod and namespace labels.

cilium-operator deletes Cilium Identities when their labels are not used by any pods.

cilium-agent no longer writes to Cilium Identities.

cilium-agent only watches Cilium Identities.

pod creation (cilium-cni) is blocked on getting a matching Cilium
Identity from the watcher cache.

```release-note
Feature: Operator Manages Cilium Identities
A new feature hidden behind a flag. Disabled by default.
cilium-operator is going to manage (create and delete) Cilium Identity API objects instead of cilium-agent.
```

kind/feature

Signed-off-by: Dorde Lapcevic <dordel@google.com>
